File: Core/Model/Impl/Loc.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.IO;
using System.Xml.Serialization;
using BinkyRailways.Core.Storage;
using BinkyRailways.Core.Util;
using Newtonsoft.Json;

namespace BinkyRailways.Core.Model.Impl
{
    /// <summary>
    /// Single locomotive.
    /// </summary>
    [XmlRoot]
    public class Loc : PersistentEntity, ILoc
    {
        private const string ImageId = "Image";


        private readonly Property<Address> address;
        private readonly Property<int> speedSteps;
        private readonly Property<int> slowSpeed;
        private readonly Property<int> mediumSpeed;
        private readonly Property<int> maximumSpeed;
        private readonly Property<ChangeDirection> changeDirection;
        private readonly Property<string> owner;
        private readonly Property<string> remarks;
        private readonly LocFunctions functions;

        /// <summary>
        /// Default ctor
        /// </summary>
        public Loc()
        {
            address = new Property<Address>(this, null);
            speedSteps = new Property<int>(this, DefaultValues.DefaultLocSpeedSteps);
            slowSpeed = new Property<int>(this, DefaultValues.DefaultLocSlowSpeed);
            mediumSpeed = new Property<int>(this, DefaultValues.DefaultLocMediumSpeed);
            maximumSpeed = new Property<int>(this, DefaultValues.DefaultLocMaximumSpeed);
            changeDirection = new Property<ChangeDirection>(this, DefaultValues.DefaultLocChangeDirection);
            owner = new Property<string>(this, DefaultValues.DefaultLocOwner);
            remarks = new Property<string>(this, DefaultValues.DefaultLocRemarks);
            functions = new LocFunctions(this);
        }

        /// <summary>
        /// Address of the locomotive
        /// </summary>
        [JsonIgnore]
        [XmlIgnore]
        public Address Address
        {
            get { return address.Value; }
            set { address.Value = value; }
        }

        /// <summary>
        /// Gets all (non-null) addresses configured in this entity.
        /// </summary>
        [JsonIgnore]
        [XmlIgnore]
        public virtual IEnumerable<AddressUsage> AddressUsages
        {
            get
            {
                var addr = Address;
                if (addr != null) yield return new AddressUsage(addr, AddressDirection.Output);
            }
        }

        /// <summary>
        /// Percentage of speed steps for the slowest speed of this loc.
        /// Value between 1 and 100.
        /// </summary>
        [DefaultValue(DefaultValues.DefaultLocSlowSpeed)]
        public int SlowSpeed
        {
            get { return slowSpeed.Value; }
            set { slowSpeed.Value = LimitSpeedValue(value); }
        }

        /// <summary>
        /// Percentage of speed steps for the medium speed of this loc.
        /// Value between 1 and 100.
        /// </summary>
        [DefaultValue(DefaultValues.DefaultLocMediumSpeed)]
        public int MediumSpeed
        {
            get { return mediumSpeed.Value; }
            set { mediumSpeed.Value = LimitSpeedValue(value); }
        }

        /// <summary>
        /// Percentage of speed steps for the maximum speed of this loc.
        /// Value between 1 and 100.
        /// </summary>
        [DefaultValue(DefaultValues.DefaultLocMaximumSpeed)]
        public int MaximumSpeed
        {
            get { return maximumSpeed.Value; }
            set { maximumSpeed.Value = LimitSpeedValue(value); }
        }

        /// <summary>
        /// Gets the number of speed steps supported by this loc.
        /// </summary>
        [DefaultValue(DefaultValues.DefaultLocSpeedSteps)]
        public int SpeedSteps
        {
            get { return speedSteps.Value; }
            set { speedSteps.Value = value; }
        }

        /// <summary>
        /// Gets/sets the image of the given loc.
        /// </summary>
        /// <value>Null if there is no image.</value>
        /// <remarks>Image must be png, bmp, gif, jpg, wmf or emf</remarks>
        [JsonIgnore]
        [XmlIgnore]
        public Stream Image
        {
            get 
            { 
                var pkg = Package;
                return (pkg != null) ? pkg.GetGenericPart(this, ImageId) : null;
            }
            set
            {
                var pkg = Package;
                if (pkg == null)
                {
                    throw new ArgumentException("Cannot set image when not part of a package");
                }
                if (value == null)
                {
                    pkg.RemoveGenericPart(this, ImageId);
                }
                else
                {
                    pkg.SetGenericPart(this, ImageId, value);
                }
            }
        }

        /// <summary>
        /// Is it allowed for this loc to change direction?
        /// </summary>
        [DefaultValue(DefaultValues.DefaultLocChangeDirection)]
        public ChangeDirection ChangeDirection
        {
            get { return changeDirection.Value; }
            set { changeDirection.Value = value; }
        }

        /// <summary>
        /// Gets the names of all functions supported by this loc.
        /// </summary>
        public LocFunctions Functions
        {
            get { return functions; }
        }

        /// <summary>
        /// Gets the names of all functions supported by this loc.
        /// </summary>
        ILocFunctions ILoc.Functions
        {
            get { return Functions; }
        }

        /// <summary>
        /// Gets/sets the name of the person that owns this loc.
        /// </summary>
        [DefaultValue(DefaultValues.DefaultLocOwner)]
        public string Owner
        {
            get { return owner.Value; }
            set { owner.Value = value ?? string.Empty; }
        }

        /// <summary>
        /// Gets/sets remarks (free text) about this loc.
        /// </summary>
        [DefaultValue(DefaultValues.DefaultLocRemarks)]
        public string Remarks
        {
            get { return remarks.Value; }
            set { remarks.Value = value ?? string.Empty; }
        }

        /// <summary>
        /// Address of the locomotive.
        /// Used for serialization.
        /// </summary>
        [JsonProperty("Address")]
        [XmlElement("Address")]
        [Browsable(false)]
        [EditorBrowsable(EditorBrowsableState.Never)]
        public string AddressString
        {
            get { return Default<AddressTypeConverter>.Instance.ConvertToString(Address); }
            set { Address = (Address) Default<AddressTypeConverter>.Instance.ConvertFromString(value); }
        }

        /// <summary>
        /// Accept a visit by the given visitor
        /// </summary>
        public override TReturn Accept<TReturn, TData>(EntityVisitor<TReturn, TData> visitor, TData data)
        {
            return visitor.Visit(this, data);
        }

        /// <summary>
        /// Validate the integrity of this entity.
        /// </summary>
        public override void Validate(IEntity validationRoot, ValidationResults results)
        {
            base.Validate(validationRoot, results);
            if (Address == null)
            {
                results.Warn(this, Strings.WarnNoAddressSpecified);
            }
            var slow = SlowSpeed;
            var medium = MediumSpeed;
            var max = MaximumSpeed;
            if (slow > medium) 
                results.Warn(this, Strings.WarnSlowSpeedHigherThenMedium);
            if (max < medium) 
                results.Warn(this, Strings.WarnMaxSpeedLowerThenMedium);
        }

        /// <summary>
        /// Human readable name of this type of entity.
        /// </summary>
        public override string TypeName
        {
            get { return Strings.TypeNameLoc; }
        }

        /// <summary>
        /// Gets package relative folder for this type of entity.
        /// </summary>
        internal override string PackageFolder
        {
            get { return PackageFolders.Loc; }
        }

        /// <summary>
        /// Compare the last modification of this entity (from the import source) with the given entity found in
        /// the target package.
        /// </summary>
        /// <param name="target">The equal entity in the target package. Can be null.</param>
        ImportComparison IImportableEntity.CompareTo(IImportableEntity target)
        {
            return CompareTo((IPersistentEntity)target);
        }

        /// <summary>
        /// Import this entity into the given package.
        /// </summary>
        void IImportableEntity.Import(IPackage target)
        {
            ((Package)target).Import(this);
            target.Railway.Locs.Add(target.GetLoc(Id));
        }

        /// <summary>
        /// Limit the given value such that is falls between 1 and 100
        /// </summary>
        private static int LimitSpeedValue(int speed)
        {
            return Math.Max(1, Math.Min(speed, 100));
        }
    }
}

File: pkg/core/model/impl/loc.go
// Copyright 2020 Ewout Prangsma
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author Ewout Prangsma
//

package impl

import (
	"context"
	"iter"

	"github.com/binkyrailways/BinkyRailways/pkg/core/model"
)

// Loc extends model Loc with implementation methods
type Loc interface {
	Entity
	model.Loc
	PersistentEntity
}

type loc struct {
	entity
	persistentEntity

	Address         model.Address
	SlowSpeed       int
	MediumSpeed     int
	MaximumSpeed    int
	SpeedSteps      int
	ChangeDirection model.ChangeDirection
	Owner           string
	Remarks         string
	VehicleType     model.VehicleType
	Functions       locFunctions `xml:"Functions"`
}

var (
	_ Loc = &loc{}
)

const (
	imageID = "Image"
)

// NewLoc initialize a new loc
func NewLoc(p model.Package) Loc {
	l := &loc{
		SlowSpeed:       model.DefaultLocSlowSpeed,
		MediumSpeed:     model.DefaultLocMediumSpeed,
		MaximumSpeed:    model.DefaultLocMaximumSpeed,
		SpeedSteps:      model.DefaultLocSpeedSteps,
		ChangeDirection: model.DefaultLocChangeDirection,
		Owner:           model.DefaultLocOwner,
		Remarks:         model.DefaultLocRemarks,
		VehicleType:     model.DefaultLocVehicleType,
	}
	l.EnsureID()
	l.SetPackage(p)
	l.SetDescription("New loc")
	l.persistentEntity.Initialize(l.entity.OnModified)
	l.Functions.SetContainer(l)
	return l
}

// GetEntityType returns the type of this entity
func (l *loc) GetEntityType() string {
	return TypeLoc
}

// Accept a visit by the given visitor
func (l *loc) Accept(v model.EntityVisitor) interface{} {
	return v.VisitLoc(l)
}

// ForEachAddress iterates all addresses in this entity and any child entities.
func (l *loc) ForEachAddress(cb func(addr model.Address, onUpdate func(context.Context, model.Address) error)) {
	cb(l.Address, l.SetAddress)
}

// Get the Address of the entity
func (l *loc) GetAddress() model.Address {
	return l.Address
}

// Set the Address of the entity
func (l *loc) SetAddress(ctx context.Context, value model.Address) error {
	if !l.Address.Equals(value) {
		l.Address = value
		l.OnModified()
	}
	return nil
}

// Return a sequence of all (non-empty) addresses configured in this
// entity with the direction their being used.
func (l *loc) AllAddressUsages() iter.Seq[model.AddressUsage] {
	return func(yield func(model.AddressUsage) bool) {
		if !l.Address.IsEmpty() {
			yield(model.AddressUsage{
				Address:   l.Address,
				Direction: model.AddressDirectionOutput,
			})
		}
	}
}

// Get percentage of speed steps for the slowest speed of this loc.
// Value between 1 and 100.
func (l *loc) GetSlowSpeed() int {
	return l.SlowSpeed
}
func (l *loc) SetSlowSpeed(value int) error {
	if l.SlowSpeed != value {
		l.SlowSpeed = value
		l.OnModified()
	}
	return nil
}

// Get percentage of speed steps for the medium speed of this loc.
// Value between 1 and 100.
func (l *loc) GetMediumSpeed() int {
	return l.MediumSpeed
}

// Set percentage of speed steps for the medium speed of this loc.
// Value between 1 and 100.
func (l *loc) SetMediumSpeed(value int) error {
	if l.MediumSpeed != value {
		l.MediumSpeed = value
		l.OnModified()
	}
	return nil
}

// Get percentage of speed steps for the maximum speed of this loc.
// Value between 1 and 100.
func (l *loc) GetMaximumSpeed() int {
	return l.MaximumSpeed
}

// Set percentage of speed steps for the maximum speed of this loc.
// Value between 1 and 100.
func (l *loc) SetMaximumSpeed(value int) error {
	if l.MaximumSpeed != value {
		l.MaximumSpeed = value
		l.OnModified()
	}
	return nil
}

// Gets the number of speed steps supported by this loc.
func (l *loc) GetSpeedSteps() int {
	return l.SpeedSteps
}

// Sets the number of speed steps supported by this loc.
func (l *loc) SetSpeedSteps(value int) error {
	if l.SpeedSteps != value {
		l.SpeedSteps = value
		l.OnModified()
	}
	return nil
}

// Gets the image (content) of this loc
func (l *loc) GetImage() []byte {
	img, _ := l.GetPackage().GetGenericPart(l, imageID)
	return img
}

// Sets the image (content) of this loc
// Image must be png, bmp, gif, jpg, wmf or emf.
func (l *loc) SetImage(value []byte) error {
	if len(value) == 0 {
		return l.GetPackage().RemoveGenericPart(l, imageID)
	} else {
		return l.GetPackage().SetGenericPart(l, imageID, value)
	}
}

// Get if it is allowed for this loc to change direction.
func (l *loc) GetChangeDirection() model.ChangeDirection {
	return l.ChangeDirection
}

// Set if it is allowed for this loc to change direction.
func (l *loc) SetChangeDirection(value model.ChangeDirection) error {
	if l.ChangeDirection != value {
		l.ChangeDirection = value
		l.OnModified()
	}
	return nil
}

// Gets the names of all functions supported by this loc.
func (l *loc) GetFunctions() model.LocFunctions {
	return &l.Functions
}

// Gets the name of the person that owns this loc.
func (l *loc) GetOwner() string {
	return l.Owner
}

// Sets the name of the person that owns this loc.
func (l *loc) SetOwner(value string) error {
	if l.Owner != value {
		l.Owner = value
		l.OnModified()
	}
	return nil
}

// Gets remarks (free text) about this loc.
func (l *loc) GetRemarks() string {
	return l.Remarks
}

// Sets remarks (free text) about this loc.
func (l *loc) SetRemarks(value string) error {
	if l.Remarks != value {
		l.Remarks = value
		l.OnModified()
	}
	return nil
}

// Gets the type of vehicle this loc is.
func (l *loc) GetVehicleType() model.VehicleType {
	return l.VehicleType
}

// Sets the type of vehicle this loc is.
func (l *loc) SetVehicleType(value model.VehicleType) error {
	if l.VehicleType != value {
		l.VehicleType = value
		l.OnModified()
	}
	return nil
}

// Upgrade to latest version
func (l *loc) Upgrade() {
	if l.VehicleType == "" {
		l.VehicleType = model.DefaultLocVehicleType
	}
}

// GetRailway returns the containing railway
func (l *loc) GetRailway() model.Railway {
	if pkg := l.GetPackage(); pkg != nil {
		return pkg.GetRailway()
	}
	return nil
}
